### PR TITLE
Updates SDR classifier internals

### LINKED ISSUE #3221: Update SDRClassifier internals
When [this PR](https://github.com/numenta/nupic.core/pull/1010) is merged, the capnp proto for SDR Classifiers will change. 

Thus its implementation should be updated, as well as the corresponding tests if necessary.

### PR DESCRIPTION
**Pull request by natoromano**
Fixes #3221 et #3222, respectively by updating the SDR classifier internals to match the new serialization proto for SDR classifier, and by fixing the fact that maxSteps was not serialized.

Also:
- makes C++ the default implementation for the SDR classifier
- adds a test for the C++ implementation within a SDRClassifierRegion

@mrcslws @scottpurdy 

Note: the CI tests will break until numenta/nupic.core#1010 is merged.

### CODE REVIEW

**Comment by mrcslws** [comment]
👍  Thanks for doing this. (cc: @ywcui1990)

**Comment by ywcui1990** [comment]
@natoromano  Thanks for fixing the SDRClassifier!

**Comment by scottpurdy** [comment]
👍  once numenta/nupic.core#1010 is merged

**Comment by rhyolight** [comment]
@natoromano If this PR depends on changes made in https://github.com/numenta/nupic.core/pull/1010 you'll need to update the SHA in `.nupic_modules` to `e71b815c6c044918f4d58e12e169a2376b41845d` in this PR.

**Comment by natoromano** [comment]
@scottpurdy should I use this PR to also update the SDRClassifierRegion to handle `cpp` as implementation? And everything that goes along with that?

**Comment by scottpurdy** [comment]
@natoromano whichever is easier for you!

**Comment by scottpurdy** [comment]
@natoromano let me know if you want me to merge this!

**Comment by scottpurdy** [comment]
@natoromano looks like valid Travis failure for something related to versions.

**Comment by natoromano** [comment]
There appears to be a slight discrepancy when re-training a classifier. I'll look into it.

File: src/nupic/regions/SDRClassifierRegion.py [hunk 1]
```diff
@@ -385,7 +385,7 @@ def compute(self, inputs, outputs):
                                     infer=False)
 
     # fill outputs with clResults
-    if clResults is not None:
+    if clResults is not None and clResults:
```
Comment (scottpurdy) [inline src/nupic/regions/SDRClassifierRegion.py]: What's the intent of the addition here? Are you trying to make sure that `clResults` is not an empty dict? It's better to check that explicitly.

**Comment by natoromano** [comment]
@scottpurdy ready!

**Comment by rhyolight** [comment]
@natoromano said:
> Fixes #3221, fixes #3222

For future reference, I think they need to be on different lines:

```
fixes #123
fixes #345
```

**Pull request merged**